src/iso19111/operation/singleoperation.cpp

```diff
@@ -384,6 +407,92 @@ CoordinateOperationNNPtr CoordinateOperation::shallowClone() const {
+
+// ---------------------------------------------------------------------------
+
+//! @cond Doxygen_Suppress
+CoordinateTransformer::~CoordinateTransformer() {
+    if (d->pj_) {
+        proj_assign_context(d->pj_, pj_get_default_ctx());
+        proj_destroy(d->pj_);
+    }
+}
+//! @endcond
+
+// ---------------------------------------------------------------------------
+
+CoordinateTransformerNNPtr
+CoordinateTransformer::create(const CoordinateOperationNNPtr &op,
+                              PJ_CONTEXT *ctx) {
+    auto transformer = NN_NO_CHECK(
+        CoordinateTransformer::make_unique<CoordinateTransformer>());
+    transformer->d->pj_ = pj_obj_create(ctx, op);
+    if (transformer->d->pj_ == nullptr)
+        throw util::UnsupportedOperationException(
+            "Cannot instantiate transformer");
+    return transformer;
+}
+
+// ---------------------------------------------------------------------------
+
+/** Transforms a coordinate tuple.
+ *
+ * PJ_COORD is a union of many structures. In the context of this method,
+ * it is prudent to only use the v[] array, with the understanding that
+ * the expected input values should be passed in the order and the unit of
+ * the successive axis of the input CRS. Similarly the values returned in the
+ * v[] array of the output PJ_COORD are in the order and the unit of the
+ * successive axis of the output CRS.
+ * For coordinate operations involving a time-dependent operation,
+ * coord.v[3] is the decimal year of the coordinate epoch of the input (or
+ * HUGE_VAL to indicate none)
+ *
+ * If an error occurs, HUGE_VAL is returned in the .v[0] member of the output
+ * coordinate tuple.
+ *
+ * Example how to transform coordinates from EPSG:4326 (WGS 84
+ * latitude/longitude) to EPSG:32631 (WGS 84 / UTM zone 31N).
+\code{.cpp}
+    auto authFactory =
+        AuthorityFactory::create(DatabaseContext::create(), std::string());
+    auto coord_op_ctxt = CoordinateOperationContext::create(
+        authFactory, nullptr, 0.0);
+    auto authFactoryEPSG =
+        AuthorityFactory::create(DatabaseContext::create(), "EPSG");
+    auto list = CoordinateOperationFactory::create()->createOperations(
+        authFactoryEPSG->createCoordinateReferenceSystem("4326"),
+        authFactoryEPSG->createCoordinateReferenceSystem("32631"),
+        coord_op_ctxt);
+    ASSERT_TRUE(!list.empty());
+    PJ_CONTEXT* ctx = proj_context_create();
+    auto transformer = list[0]->coordinateTransformer(ctx);
+    PJ_COORD c;
+    c.v[0] = 49; // latitude in degree
+    c.v[1] = 2;  // longitude in degree
+    c.v[2] = 0;
+    c.v[3] = HUGE_VAL;
+    c = transformer->transform(c);
+    EXPECT_NEAR(c.v[0], 426857.98771728, 1e-8); // easting in metre
+    EXPECT_NEAR(c.v[1], 5427937.52346492, 1e-8); // northing in metre
+    proj_context_destroy(ctx);
+\endcode
+ */
+PJ_COORD CoordinateTransformer::transform(PJ_COORD coord) {
+    return proj_trans(d->pj_, PJ_FWD, coord);
+}
```

Comment:
Sorry for late feedback after this is already merged. Would be nice to be able to pass `proj_direction` as parameter.

Reply:
> Would be nice to be able to pass `proj_direction` as parameter.

you can call inverse() on a CoordinateOperation